include/exadg/time_integration/restart.h

```diff
@@ -138,6 +155,739 @@ read_write_distributed_vector(VectorType & vector, BoostArchiveType & archive)
+      ++cell_batch_idx)
+  {
+    fe_eval.reinit(cell_batch_idx);
+    for(const unsigned int q : fe_eval.quadrature_point_indices())
+    {
+      dealii::Point<dim, dealii::VectorizedArray<Number>> const cell_batch_points =
+        fe_eval.quadrature_point(q);
+      for(unsigned int i = 0; i < dealii::VectorizedArray<Number>::size(); ++i)
+      {
+        dealii::Point<dim> p;
+        for(unsigned int d = 0; d < dim; ++d)
+        {
+          p[d] = cell_batch_points[d][i];
+        }
+        integration_points.push_back(p);
+      }
+    }
+  }
+
+  return integration_points;
+}
+
+/**
+ * Utility function to compute the right hand side of a projection with values given in integration
+ * points obtained via `collect_integration_points()`.
+ */
+template<int dim, int n_components, typename Number, typename VectorType>
+inline VectorType
+assemble_projection_rhs(
+  dealii::MatrixFree<dim, Number, dealii::VectorizedArray<Number>> const & matrix_free,
+  CellIntegrator<dim, n_components, Number> &                              fe_eval,
+  std::vector<
+    typename dealii::FEPointEvaluation<n_components, dim, dim, Number>::value_type> const &
+                     values_source_in_q_points_target,
+  unsigned int const dof_index)
+{
+  VectorType system_rhs;
+  matrix_free.initialize_dof_vector(system_rhs, dof_index);
+
+  unsigned int idx_q_point = 0;
+
+  for(unsigned int cell_batch_idx = 0; cell_batch_idx < matrix_free.n_cell_batches();
+      ++cell_batch_idx)
+  {
+    fe_eval.reinit(cell_batch_idx);
+    for(unsigned int const q : fe_eval.quadrature_point_indices())
+    {
+      dealii::Tensor<1, n_components, dealii::VectorizedArray<Number>> tmp;
+
+      for(unsigned int i = 0; i < dealii::VectorizedArray<Number>::size(); ++i)
+      {
+        typename dealii::FEPointEvaluation<n_components, dim, dim, Number>::value_type const
+          values = values_source_in_q_points_target[idx_q_point];
+
+        // Increment index into `values_source_in_q_points_target` which is dictated
+        // by `integration_points_target`, i.e., `collect_integration_points()`.
+        ++idx_q_point;
+
+        if constexpr(n_components == 1)
+        {
+          tmp[0][i] = values;
+        }
+        else
+        {
+          for(unsigned int d = 0; d < n_components; ++d)
+          {
+            tmp[d][i] = values[d];
+          }
+        }
+      }
+
+      fe_eval.submit_value(tmp, q);
+    }
+    fe_eval.integrate(dealii::EvaluationFlags::values);
+    fe_eval.distribute_local_to_global(system_rhs);
+  }
+  system_rhs.compress(dealii::VectorOperation::add);
+
+  return system_rhs;
+}
+
+/**
+ * Utilitiy function to project vectors from a source to a target triangulation
+ * via `dealii::RemotePointEvaluation`, matrix-free mass operator evaluation
+ * and a Jacobi-preconditioned CG solver.
+ */
+template<int dim, typename Number, int n_components, typename VectorType>
+inline void
+project_vectors(
```

Comment:
Everything in this file related to the projection should not be in `time_integration/restart.h` in my opinion, but in `include/exadg/operators` (where we have e.g. utility functions for interpolation between solutions on different grids, see https://github.com/exadg/exadg/blob/master/include/exadg/operators/solution_interpolation_between_triangulations.h). Right now, I would argue such functionality does not belong to `include/exadg/grid`, because finite elements and dof handlers are involved here (not just triangulations and mapping), hence `include/exadg/operators`.

Reply:
good, I will split off the projection introduced here and put it into `include/exadg/operators/solution_projection_between_triangulations.h`
I tried to use the functions in there actually first, but it did not work for Raviart-Thomas for some reason so I opted for the projection and `RemotePointEvaluation`.
For the current use case -- I think it is `FE_Q` and `FE_DGQ` only, grid-to-grid interpolation works. If I grep `SolutionInterpolationBetweenTriangulations`, I get zero hits in ExaDG, so I did not bother with this further although I would say the `ElementType` has to be asserted within `SolutionInterpolationBetweenTriangulations`. I think @jh66637 is using this so I just left it there ;)

Reply:
I am currently not using it anymore, there are more advanced interpolation methods I am relying on, which I want to move to ExaDG when I find the time. I would still like to keep it. It is quite handy to have multiple options how to interpolate between grids :)